File: lib/rb/test/fuzz/fuzz_common.rb
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.
#

$:.unshift File.expand_path('../../lib', __dir__)
$:.unshift File.expand_path('../../ext', __dir__)
require 'thrift'
$:.unshift File.dirname(__FILE__) + "/gen-rb"
require 'fuzz_test_constants'

require 'coverage'
Coverage.start(branches: true) unless Coverage.respond_to?(:running?) && Coverage.running?
require 'ruzzy'
# Ruzzy.enable_branch_coverage_hooks

def ignorable_fuzz_exception?(error)
  return true if error.is_a?(Thrift::ProtocolException) ||
    error.is_a?(EOFError) ||
    error.is_a?(Encoding::UndefinedConversionError)

  [
    /don't know what (?:c)?type/,
    /Too many fields for union/,
    /too big to convert to '(?:int|long)'/,
    /bignum too big to convert into 'long'/,
    /negative array size/,
    /Union fields are not set/
  ].any? { |pattern| error.message =~ pattern }
end

def read_fuzz_test(protocol, read_message_begin)
  protocol.read_message_begin if read_message_begin
  obj = Fuzz::FuzzTest.new
  obj.read(protocol)
  protocol.read_message_end if read_message_begin
  obj
end

def write_fuzz_test(protocol, obj, write_message_begin)
  if write_message_begin
    protocol.write_message_begin('fuzz', Thrift::MessageTypes::CALL, 0)
  end
  obj.write(protocol)
  protocol.write_message_end if write_message_begin
end

def create_parser_fuzzer(protocol_factory_class, read_message_begin: false)
  lambda do |data|
    transport = Thrift::MemoryBufferTransport.new(data)
    protocol = protocol_factory_class.new.get_protocol(transport)
    read_fuzz_test(protocol, read_message_begin)
    0
  rescue StandardError => e
    # We're looking for memory corruption, not Ruby exceptions
    raise unless ignorable_fuzz_exception?(e)
  end
end

def create_roundtrip_fuzzer(protocol_factory_class, read_message_begin: false)
  lambda do |data|
    transport = Thrift::MemoryBufferTransport.new(data)
    protocol = protocol_factory_class.new.get_protocol(transport)
    obj = read_fuzz_test(protocol, read_message_begin)

    serialized_data = +""
    transport2 = Thrift::MemoryBufferTransport.new(serialized_data)
    protocol2 = protocol_factory_class.new.get_protocol(transport2)
    write_fuzz_test(protocol2, obj, read_message_begin)

    transport3 = Thrift::MemoryBufferTransport.new(serialized_data)
    protocol3 = protocol_factory_class.new.get_protocol(transport3)
    deserialized_obj = read_fuzz_test(protocol3, read_message_begin)

    raise "Roundtrip mismatch" unless obj == deserialized_obj
    0
  rescue StandardError => e
    # We're looking for memory corruption, not Ruby exceptions
    raise unless ignorable_fuzz_exception?(e)
  end
end

File: lib/nodejs/test/fuzz/fuzz_common.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

const thrift = require("../../lib/thrift");
const FuzzTest = require("./gen-nodejs/FuzzTestNoUuid_types");
// const { FuzzedDataProvider } = require("@jazzer.js/core");

/**
 * Creates a parser fuzzer function for a specific protocol
 * @param {Function} protocolFactory - The Thrift protocol factory function
 * @param {boolean} [readMessageBegin=false] - Whether to call readMessageBegin before reading the test instance
 * This is needed for protocols that do not support parsing just a struct, such as TJSONProtocol.
 * @returns {Function} A fuzzer function that can be used with Jazzer.js
 */
function createParserFuzzer(protocolFactory, readMessageBegin = false) {
  return function fuzz(data) {
    if (data.length < 2) {
      return;
    }

    try {
      // Set up transport with input data
      const transport = new thrift.TFramedTransport(data);
      const protocol = protocolFactory(transport);
      const testInstance = new FuzzTest.FuzzTest();
      if (readMessageBegin) {
        protocol.readMessageBegin();
      }
      testInstance[Symbol.for("read")](protocol);
    } catch (e) {
      if (
        !(
          e.name === "InputBufferUnderrunError" ||
          e.name === "TProtocolException"
        )
      ) {
        // TODO: Are other exceptions expected?
        // console.log(e);
      }
    }
  };
}

/**
 * Creates a roundtrip fuzzer function for a specific protocol
 * @param {Function} protocolFactory - The Thrift protocol factory function
 * @param {boolean} [readMessageBegin=false] - Whether to call readMessageBegin before reading the test instance
 * This is needed for protocols that do not support parsing just a struct, such as TJSONProtocol.
 * @returns {Function} A fuzzer function that can be used with Jazzer.js
 */
function createRoundtripFuzzer(protocolFactory, readMessageBegin = false) {
  return function fuzz(data) {
    if (data.length < 2) {
      return;
    }

    try {
      // First deserialize using framed transport for input
      const transport1 = new thrift.TFramedTransport(data);
      const protocol1 = protocolFactory(transport1);
      const testInstance = new FuzzTest.FuzzTest();
      if (readMessageBegin) {
        protocol1.readMessageBegin();
      }
      testInstance[Symbol.for("read")](protocol1);

      // Then serialize using buffered transport with callback
      let serializedData;
      const transport2 = new thrift.TBufferedTransport(null, function (buf) {
        serializedData = buf;
      });
      const protocol2 = protocolFactory(transport2);
      testInstance[Symbol.for("write")](protocol2);
      protocol2.flush();

      if (!serializedData) {
        throw new Error("Serialization failed - no data produced");
      }

      // Finally deserialize again and compare using framed transport
      const transport3 = new thrift.TFramedTransport(serializedData);
      const protocol3 = protocolFactory(transport3);
      const deserialized = new FuzzTest.FuzzTest();
      if (readMessageBegin) {
        protocol3.readMessageBegin();
      }
      deserialized[Symbol.for("read")](protocol3);

      // Compare the objects
      if (!testInstance.equals(deserialized)) {
        throw new Error("Roundtrip comparison failed");
      }
    } catch (e) {
      if (
        !(
          e.name === "InputBufferUnderrunError" ||
          e.name === "TProtocolException"
        )
      ) {
        // TODO: Are other exceptions expected?
        // console.log(e);
      }
    }
  };
}

module.exports = {
  createParserFuzzer,
  createRoundtripFuzzer,
};
